### ISSUE TITLE
Incremental builds brain dump

### ISSUE THREAD

**Issue by briaguya** [cityofaustin/techstack#3152]
In order to support incremental builds, we need to think through how we're exposing our data via the Joplin API as well as how we're consuming it in Janis.

To support this, I would propose we:
## In Joplin:
- Make a new endpoint in our Joplin API specifically to get page structure
  - This could be updated in the future to only return the parts of the page structure that matter for the incremental build we're doing.
- Make a new endpoint in our Joplin API to get an "instance" of a page. This would include the parent of a page at any given url.
## In Janis:
- Update the build process to:
  - Query the page structure endpoint
    - this seems like it might get recursive, is that something we can do with graphql? what are our other options?
  - Parse the page structure and use it to make queries for the rest of the pages by id
    - sidenote: might be able to make the page specific queries inside our getData async methods in static.config.js, not sure how different this is in react static 7.

### Thoughts in scribbled form:

![IMG_3669](https://user-images.githubusercontent.com/6384156/66493320-72d5c500-ea7b-11e9-82f9-97b0a826ea47.jpg)

**Comment by briaguya** [comment]
did a bit of playing around here https://github.com/cityofaustin/joplin/pull/406

**Comment by desigonz** [comment]
This is brilliant 🤯Let's touch base closer to the end of sprint 40 about how we can start breaking this down into issues

**Comment by briaguya** [comment]
Did some POC stuff:
Joplin PR: https://github.com/cityofaustin/joplin/pull/406
Joplin app: http://joplin-pr-3151-urls-in-api.herokuapp.com
Janis PR: https://github.com/cityofaustin/janis/pull/559
Janis app: https://deploy-preview-559--janis.netlify.com/

Overview:
1. Joplin does some magic to create a site structure, which is just a list of url/page id pairs + some bonus junk I'm not using yet
2. Janis requests this, then goes through and builds pages at the Joplin supplied URLs by requesting the page data

Lots of things are broken, there's a ton of copypasta, but it seems like this could work. (I've proved the concept to myself at least)

**Comment by briaguya** [comment]
@desigonz I think it makes sense to turn this into an epic. I'll brain dump here first:

Epic "Implement incremental builds":
- [ ] Set up epic branch staging environment. 
  - All PRs related to this epic will go into this branch until we hit MVP to go live.
  - QA will be done on this staging environment. 
    - In order to fully test this, we will need to be able to:
      - trigger builds 
      - generate previews
  - related: @Niicck! Let's talk https://github.com/cityofaustin/techstack/issues/2969 soon!
- [ ] Implement "incremental ready" build process
  - Joplin should expose an API endpoint for:
    - The information architecture/url structure of the site
    - Each page by global ID with enough information returned to build the whole page, this includes all children/top links/contextual nav info etc.
      - These can vary by url, How/where we want to handle that? Can we build some of it into JanisBasePage?
  - Janis should:
    - Request a list of pages to build, and get a list of url/id pairs.
    - Build a page at each url by requesting data by id
  - Once this is complete, a release branch should be created off of the main incremental builds epic branch. QA can then be performed on this release branch, and once everything passes, we can merge to staging/prod. Even though this isn't incremental yet, it is quite possible we can speed up builds with better parallelization.
- [ ] Upgrade janis to "incremental ready" static site generator
  - This will probably be react-static 7, but we have options.
- [ ] Implement proof-of-concept-manually-triggerable-on-a-dev-machine (POCMTODM - how's that for alphabet soup?) incremental builds.
  - bonus points: make a button in a pr app to magic demo this, if we have https://github.com/cityofaustin/techstack/issues/2969 done it could probably be faked for a demo.
- [ ] Create joplin endpoint for incremental builds
  - How do we want to build this? Do we need to? We've got IDs? Should the trigger send the ids?
- [ ] Create joplin trigger for incremental builds
  - Right now every time we update something we tell Janis "rebuild!," we need to say "rebuild x and y, delete z, add w, etc." do we want that to be in the trigger? How do we want to implement this?
- [ ] Implement QA-able incremental builds
  - Need acceptance criteria for this, but basically on publish/unpublish/update/edit/all the things content authors do we need to add/update/remove all the right things in all the right places.
  - Need to define MVP for this, eg. unpublishing - MVP could be "rebuild everything", same for add, we could have it so it's just updating a page that works incrementally at first and work from there.
- Keep making it better!

**Comment by Niicck** [comment]
Thoughts
- Re: "Implement QA-able incremental builds"
  - Here is a great opportunity for us to start doing unit tests. When we delete a page in Joplin, does it delete it from Janis? When we add a page, does it add a page to Janis? When 3 users hit a trigger at the same time, will it properly queue those requests? etc...
- Re: "Create joplin endpoint for incremental builds"
  - This can tie into the new build process that I'm working on. Passing in ids/info to the incremental build trigger is essential.
- Re: "Upgrade janis to "incremental ready" static site generator"
  - I vote react-static 7 at this point. Gatsby doesn't do incremental builds.
  - An alternative is using a Server-side rendering framework like next.js. We'd no longer be building a static site, but its one way to handle builds/navigation. Now would be the time to ponder other options if we were ever going to do it.
- Re: "Set up epic branch staging environment"
  - Step 1: update what we've already got to react-static 7. This is an incremental change we can do for our current Janis. I agree on the point: "All PRs related to this epic will go into this branch until we hit MVP to go live." However, the fewer PRs that go into that fork/branch, the easier it will be for us to merge/maintain. I think the update to react-static 7 could/should be done and merged into staging/prod first. That way (hopefully) the epic fork/branch changes will be confined to static.config.js + helpers as opposed to the entire frontend.

**Comment by Niicck** [comment]
Also, I'd like to get an in-person pitch from Brian as to how he envisions the requirements for incremental builds. There are things my brain is missing.

**Comment by Niicck** [comment]
Additional Thought about entertaining Server-side rendering.

- Search
  - Is a feature that is coming up soon. This is going to be massive and important. Would server-side rendering be a better approach for supporting search? 

Thinking out loud. I'm doing more reading.

**Comment by desigonz** [comment]
💖to all of this! How does an in-person pitch from @briaguya at the end of Monday's dev sync sound? Or we could set up a separate meeting for this.

**Comment by briaguya** [comment]
@Niicck re: Re: "Set up epic branch staging environment"

mostly this is for making "Implement "incremental ready" build process" happen as this can be done without moving to react-static 7.

basically it's a "let's have a good place to test those PRs I made as we flesh them out as best we can" requirement

we can discuss more in person

**Comment by easherma** [comment]
re: search; there are plenty of ways to do that on a static site
extra thought from other thoughts: as part of setting up this infrastructure, we could move some of the logic out of janis and into the publishing infrastructure. thinking specifically about the cleanData stuff and other middleware-esque things that are hard to debug in javascript and would be nice to have more control over what happens depending on what kind of error

**Comment by easherma** [comment]
We should also (perhaps mostly for @desigonz but also so we keep our heads screwed on straight and have a record) document WHY we want to do this and the value we expect to gain, so we have some goals to check progress against. 

For instance, if the goal is to save build time, we'll also probably want to look at how long the webpack compiling takes and if that can be optimized. 

But that could just be one of several goals.

**Comment by briaguya** [comment]
@Easherma incremental scales like mad woah, other things should happen too but building 1 page vs 1000 pages is gonna need to happen.

**Comment by briaguya** [comment]
@easherma re: not using cleanData all over the place:
```
Joplin should expose an API endpoint for:
Each page by global ID with enough information returned to build the whole page, this includes all children/top links/contextual nav info etc.
These can vary by url, How/where we want to handle that? Can we build some of it into JanisBasePage?
```

**Comment by desigonz** [comment]
Awesome, thanks all. @briaguya documented your checkboxes in the incremental builds epic (#3001). Queued up #3188 for next sprint. Closing this issue :)

**Comment by easherma** [comment]
There is some good info/discussion here.
